src/main/scala/devices/i2c/I2C.scala

```diff
@@ -0,0 +1,515 @@
+
+  when (arbLost) {
+    bitCmd      := I2C_CMD_NOP
+    transmitBit := false.B
+    shift       := false.B
+    load        := false.B
+    cmdAck      := false.B
+    byteState   := s_byte_idle
+    receivedAck := false.B
+  }
+  .otherwise {
+    transmitBit := receivedData(7)
+    shift       := false.B
+    load        := false.B
+    cmdAck      := false.B
+
+    switch (byteState) {
+      is (s_byte_idle) {
+        when (go) {
+          when (cmd.start) {
+            byteState := s_byte_start
+            bitCmd    := I2C_CMD_START
+          }
+          .elsewhen (cmd.read) {
+            byteState := s_byte_read
+            bitCmd    := I2C_CMD_READ
+          }
+          .elsewhen (cmd.write) {
+            byteState := s_byte_write
+            bitCmd    := I2C_CMD_WRITE
+          }
+          .otherwise { // stop
+            byteState := s_byte_stop
+            bitCmd    := I2C_CMD_STOP
+          }
+
+          load        := true.B
+        }
+      }
+      is (s_byte_start) {
+        when (bitCmdAck) {
+          when (cmd.read) {
+            byteState := s_byte_read
+            bitCmd    := I2C_CMD_READ
+          }
+          .otherwise {
+            byteState := s_byte_write
+            bitCmd    := I2C_CMD_WRITE
+          }
+
+          load        := true.B
+        }
+      }
+      is (s_byte_write) {
+        when (bitCmdAck) {
+          when (bitCntDone) {
+            byteState := s_byte_ack
+            bitCmd    := I2C_CMD_READ
+          }
+          .otherwise {
+            byteState := s_byte_write
+            bitCmd    := I2C_CMD_WRITE
+            shift     := true.B
+          }
+        }
+      }
+      is (s_byte_read) {
+        when (bitCmdAck) {
+          when (bitCntDone) {
+            byteState := s_byte_ack
+            bitCmd    := I2C_CMD_WRITE
+          }
+          .otherwise {
+            byteState := s_byte_read
+            bitCmd    := I2C_CMD_READ
+          }
+
+          shift       := true.B
+          transmitBit := cmd.ack
+        }
+      }
+      is (s_byte_ack) {
+        when (bitCmdAck) {
+          when (cmd.stop) {
+            byteState := s_byte_stop
+            bitCmd    := I2C_CMD_STOP
+          }
+          .otherwise {
+            byteState := s_byte_idle
+            bitCmd    := I2C_CMD_NOP
+
+	    // generate command acknowledge signal
+            cmdAck    := true.B
+          }
+
+	  // assign ack_out output to bit_controller_rxd (contains last received bit)
+          receivedAck := receivedBit
+
+          transmitBit := true.B
+        }
+        .otherwise {
+          transmitBit := cmd.ack
+        }
+      }
+      is (s_byte_stop) {
+        when (bitCmdAck) {
+          byteState := s_byte_idle
+          bitCmd    := I2C_CMD_NOP
+
+	  // assign ack_out output to bit_controller_rxd (contains last received bit)
+          cmdAck    := true.B
+        }
+      }
+    }
+  }
+
+
+  //////// Top level ////////
+
+  // hack
```

Comment:
You suggested this code a workaround for regmap limitation and I think it is a hack.